### ISSUE TITLE
Empty placeholder rules?

### ISSUE THREAD

**Issue by evilnose** [lark-parser/lark#901]
**What is your question?**

In my grammar I sometimes want an OR-like relationship like this:

```
// allow
const var a;

// allow
const a;

// allow
var a;

// disallow
a;
```

Now, I set `keep_placeholders=True`, and it works great for most situations. But for the above situation I have to specify something like
```
stmt: CONST a
    | VAR a
    | CONST VAR a
```

which is tedious to walk. I also thought about something like this
```
stmt: CONST [VAR] a
    | [CONST] VAR a
```

which is easier to walk, but I'm not sure if this is the "right" way to do this, since it seems a bit redundant. I was wondering if there's a generally adopted approach for these situations, such as a "padder" rule, something like
```
stmt: CONST $PAD a
    | $PAD VAR a
    | CONST VAR a
```
and $PAD would automatically resolve to `None`.

**Comment by MegaIng** [comment]
No, there is currently no built in way of doing that. But I just though of a little trick: You can define this rule:

```
_pad: [/(?:b)a/]
```

This regex never matches, so this rule always just produces a `None`. You can then use it like this:

```
stmt: CONST _pad a
    | _pad VAR a
    | CONST VAR a
```

Wheter it is a good idea or not I am unsure of. It will repeatly try to match the regex and fail, so it might be a performance hindreness (especially with `parser='earley'`. I think the Lexers `!="dynamic"` will essentially optimize it away.)

Ofcourse you could always just check for correct usage at a later stage, but that is probably not in your interest.

I am however also not sure if we should implement something like this as a builtin. Maybe a NO_MATCH terminal would be interesting?

**Comment by evilnose** [comment]
I see. Now on performance, do you think something like
```
stmt: CONST [VAR] a
    | [CONST] VAR a
```
would have any considerable performance penalty?

Or I guess in my case I'm using LALR parser and contextual lexer. Do you think there would be a noticeable difference between the grammar you suggested and the one above?

And I think having a builtin NO_MATCH terminal would be great -- `keep_placeholders` is really nice, but this is the only case where it doesn't quite work.

**Comment by MegaIng** [comment]
Measure it. I don't think there will be a noticable difference when you are using LALR.

**Comment by erezsh** [comment]
> would have any considerable performance penalty?

There is no performance penalty for that in LALR. It probably won't be noticeable in Earley either.

You should also consider putting them in their own rule:

```stmt: const_var a
!const_var: CONST | [CONST] VAR
```

Or, you can use empty rules (tho there is a small performance penalty for them, it's negligible in this case).

```stmt: CONST [VAR] a
       | empty VAR a

empty:
```

But I don't know if I would recommend that.
